Verify `private/underlay/conn`

henriman opened the pull request:
Issues #14, #15 and #16 have been opened to keep track of the assumptions still present here.

jcp19 commented on private/underlay/conn/conn.go:
```diff
@@ -43,6 +43,12 @@ type Messages []ipv4.Message
 type Conn interface {
 	//@ pred Mem()
+
+	//@ ghost
+	//@ requires Mem()
+	//@ decreases
+	//@ pure IsLow() bool
```
should be a hyper func

henriman replied:
As in `pkg/addr`, I have added a TODO here to mark `IsLow` as `hyper` once Gobra issue 955 is resolved

jcp19 commented on private/underlay/conn/conn.go:
```diff
@@ -459,23 +475,43 @@ func (c *connUDPBase) Close() (err error) {
+	//@ invariant forall j int :: { m[j].GetAddr() } (0 <= j && j < i0) ==> m[j].GetAddr() == nil
+	// NOTE[henri]: Splitting up the following invariant has only marginally
+	// improved stability (if at all), thus we could consider undoing this.
+	// invariant forall j int :: { &m[j] } (i0 <= j && j < len(m)) ==> acc(&m[j])
```
I would rather undo this. Can you please open an issue in Gobra about the instability with this particular example?

henriman replied:
See [Gobra issue #957](https://github.com/viperproject/gobra/issues/957)

jcp19 commented on private/underlay/conn/conn_spec.gobra:
```diff
@@ -40,6 +40,67 @@ pred (c *connUDPBase) MemWithoutConn() {
+// NOTE[henri]: Once Gobra issue #883 is resolved, we could remove `GetClosed`
+// etc. in favor of unfolding `Mem` in the contract of `RevealIsLow` etc.
+// directly. However, I'm not sure whether that is preferrable. If it is, add
+// a corresponding TODO here.
```
Once again here, I believe it is best if we have a pure function to access a private field. Unfolding predicates in specifcations of public functions can often reveal implementation details

jcp19 commented on private/underlay/conn/conn_spec.gobra:
```diff
@@ -40,6 +40,67 @@ pred (c *connUDPBase) MemWithoutConn() {
+decreases
+pure func (c *connUDPBase) IsLow(withConn bool) bool
+
+// "Reveal the body" of `c.IsLow` (necessary bc. we can't implement `c.IsLow` yet).
+// TODO: Once Gobra issue #846 is resolved, implement this.
+ghost
+trusted
+requires p > 0
+requires  wildcard &&  withConn ==> acc(c.Mem(), _)
+requires  wildcard && !withConn ==> acc(c.MemWithoutConn(), _)
+requires !wildcard &&  withConn ==> acc(c.Mem(), p)
+requires !wildcard && !withConn ==> acc(c.MemWithoutConn(), p)
+requires c.IsLow(withConn)
+ensures   wildcard &&  withConn ==> acc(c.Mem(), _)
+ensures   wildcard && !withConn ==> acc(c.MemWithoutConn(), _)
+ensures  !wildcard &&  withConn ==> acc(c.Mem(), p)
+ensures  !wildcard && !withConn ==> acc(c.MemWithoutConn(), p)
+ensures  low(c.GetClosed(withConn)) && low(c.GetRemote(withConn) != nil)
+decreases
+func (c *connUDPBase) RevealIsLow(wildcard bool, p perm, withConn bool)
+
+// Assert `c.IsLow` (necessary bc. we can't implement `c.IsLow` yet).
+// TODO: Once Gobra issue #846 is resolved, implement this.
+ghost
+trusted
+requires  wildcard ==> acc(c.Mem(), _)
+requires !wildcard ==> acc(c.Mem(), R1)
+requires low(c.GetClosed(true)) && low(c.GetRemote(true) != nil)
+ensures   wildcard ==> acc(c.Mem(), _)
+ensures  !wildcard ==> acc(c.Mem(), R1)
+ensures  c.IsLow(true)
+decreases
+func (c *connUDPBase) AssertIsLow(wildcard bool)
```
we could implement these as hyper pure functions

henriman replied:
Will be done in PR #10

jcp19 commented on private/underlay/conn/conn_spec.gobra:
```diff
@@ -56,12 +137,54 @@ pred (c *connUDPIPv6) Mem() {
+decreases
+pure func (c *Config) Dereference() Config {
+	return unfolding c.Mem() in *c
+}
```
maybe `Deref` would be a better name

jcp19 commented on router/dataplane.go:
```diff
@@ -148,6 +148,11 @@ type bfdSession interface {
 // with the same name in private/underlay/conn/Conn.
 type BatchConn interface {
 	// @ pred Mem()
+	
+	// @ ghost
+	// @ requires Mem()
+	// @ decreases
+	// @ pure IsLow() bool
```
why do we have to state anything about the lowness of our BatchConn?

henriman commented:
@jcp19: Besides addressing your feedback, I have performed the following modifications:

- Added TODO regarding having to add `hyper` to `Conn.IsLow`, `BatchConn.IsLow`
- Updated comment on `NewReadMessages` to refer to Gobra issue 957
- Added `&& low(i) ==>` where appropriate

jcp19 commented on private/underlay/sockctrl/sockopt.go:
```diff
@@ -21,10 +21,14 @@ import (
+// @ requires  low(level) && low(opt)
+// @ preserves c.Mem() && c.IsLow()
+// @ ensures   e != nil ==> e.ErrorMem()
+// @ ensures   low(r) && low(e)
```
`e` behaves a lot like a pointer. please change this such that what is low is `e != nil`, not `e` itself. Please make sure that all similar snippets in this PR are fixed.

jcp19 commented on verification/utils/definitions/definitions.gobra:
```diff
@@ -76,6 +76,10 @@ const (
 	R54
 	R55
 	R56
+	// Some functions have parameters `wildcard bool` and `p perm`, and consume
+	// a wildcard permission amount if `wildcard` is set, and amount `p` 
+	// otherwise. If `wildcard` is set, `p` should be set to `anyPerm`.
+	anyPerm
```
I'm not sure what this means. why do you have this rather than using `R56`?

henriman replied:
My motivation was to make the code more readable, but it is rather unnecessary I suppose. I have removed it now.
